components/forms/ConstituencyFormWithSignup.tsx

```diff
@@ -0,0 +1,262 @@
+  // https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#server-side-validation-and-error-handling
+  // Was hitting into this issue: https://github.com/vercel/next.js/issues/55919
+
+  // Variables used to track state of the ConstituencyLookup components
+  const validPostcode = useRef("");
+  const [constituency, setConstituency] = useState<Constituency | null>(null);
+  const [constituencyApiLoading, setConstituencyApiLoading] = useState(false);
+
+  // Variables used to track state of the email opt-in components
+  const [formState, setFormState] = useState<FormData>(initialFormState);
+  const [emailError, setEmailError] = useState<EmailErrorCode | null>(null);
+  const formRef = useRef<HTMLFormElement | null>(null);
+
+  // Track whether the user is already subscribed via localStorage
+  const [subscribed, setSubscribed] = useState<string | null | false>(false);
+  useEffect(() => {
+    //string = subscription Date.now()
+    //null = not subscribed on client
+    //false = on server
+    setSubscribed(window.localStorage.getItem("fwd-subscribed"));
+  }, []);
+
+  const submitForm = async () => {
+    if (constituency && !formState.emailOptIn) {
+      router.push(`/constituencies/${constituency.slug}`);
+      return;
+    } else if (
+      constituency &&
+      formState.emailOptIn &&
+      formState.email &&
+      formRef.current &&
+      !formRef.current.email.validity.typeMismatch
+    ) {
+      //TODO set source codes from current url params.
+      const anResponse = await submitANForm(
+        formState.email,
+        validPostcode.current,
+        constituency,
+        process.env.NEXT_PUBLIC_AN_POSTCODE_FORM || "",
+        ["stop the tories", "movement forward", "election reminders", "join"],
+        "", // source codes,
+      );
+
+      if (anResponse.ok) {
+        window.localStorage.setItem("fwd-subscribed", Date.now().toString());
+        router.push(`/constituencies/${constituency.slug}`);
+      } else {
+        setEmailError("SERVER_ERROR"); //AN doesn't give error codes on failure
+      }
+    }
+
+    // VALIDATION
+    // Invalid email
+    if (
+      formState.emailOptIn &&
+      (!formState.email ||
+        (formRef.current && formRef.current.email.validity.typeMismatch))
+    ) {
+      setEmailError("EMAIL_INVALID");
+    }
+
+    // // no postcode or invalid postcode or constituency/address not selected
+    // if (
+    //   !validPostcode.current ||
+    //   !apiResponse ||
+    //   !apiResponse.constituencies ||
+    //   apiResponse.constituencies.length == 0
+    // ) {
+    //   // User hasn't input anything or invalid postcode
+    //   setPostError("POSTCODE_INVALID");
+    //   return;
+    // }
+
+    // //not selected constituency or address
+    // if (
+    //   apiResponse.constituencies.length > 1 &&
+    //   formState.constituencyIndex === false
+    // ) {
+    //   setPostError("UNCLEAR_CONSTITUENCY");
+    // }
```

Comment:
@jms301 , not sure these are necessary anymore, since the GO button is disabled until the constituency is known anyway - be good to get your thoughts tho?